website/blog/2025/05-05-arbiter-type-class/index.md

```diff
@@ -0,0 +1,280 @@
+
+Let's first build a simple fixed-priority arbiter that works for a specific kind of client:
+
+```scala
+class ArbiterClient extends Bundle {
+  val request = Bool()
+  val grant = Flipped(Bool())
+}
+
+class PriorityArbiter(nClients: Int) extends Module {
+  val clients = IO(Vec(nClients, Flipped(new ArbiterClient)))
+
+  val requests = clients.map(_.request)
+  val granted = PriorityEncoderOH(requests)
+
+  for ((client, grant) <- clients.zip(granted)) {
+    client.grant := grant
+  }
+}
+```
+
+<!-- truncate -->
+
+`ArbiterClient` is written from the perspective of the client--`request` is in the default orientation, intended to be driven by the client, and `grant` is flipped, intended to be driven by the arbiter.
+
+This works fine if you only ever need to arbitrate `ArbiterClient` bundles, but it doesn't generalize.
+For example, how would you use this arbiter with a `ready-valid` interface like `Decoupled`?
+
+```scala
+val needToArbitrate = IO(Vec(4, Flipped(Decoupled(UInt(8.W)))))
+val arbiter = Module(new PriorityArbiter(needToArbitrate.size))
+
+for ((client, decoupled) <- arbiter.clients.zip(needToArbitrate)) {
+  client.request := decoupled.valid
+  decoupled.ready := client.grant
+}
+```
+
+This works, but its a bit clunky.
+Considering the implementation of `PriorityArbiter`, is only 6 lines of Scala, it's not ideal that it takes another 4 lines to use it.
+
+## Generalizing the Arbiter
+
+What we really want is an arbiter that is generic to the type of the client.
+For those familiar with _Object-Oriented Programming_, you might be tempted to try to use inheritance to define a common interface for clients:
+
+```scala
+trait Arbitrable {
+  def request: Bool
+  def grant: Bool
+}
+
+class ArbiterClient extends Bundle with Arbitrable {
+  val request = Bool()
+  val grant = Flipped(Bool())
+}
+```
+
+However, for our example above, this would be a bit difficult.
+`Decoupled` is defined within Chisel itself--how can a user make Chisel's `Decoupled` inherit from `Arbitrable`?
+
+Instead, we can try something different.
+We could make the Arbiter generic to the type of the client and then use higher-order functions to extract the request and grant signals.
+
+```scala
+class GenericPriorityArbiter[A <: Data](
+    nClients: Int,
+    clientType: A
+  )(
+    requestFn: A => Bool,
+    grantFn: (A, Bool) => Unit) extends Module {
+  val clients = IO(Vec(nClients, Flipped(clientType)))
+
+  val requests = clients.map(requestFn(_))
+  val granted = PriorityEncoderOH(requests)
+
+  for ((client, grant) <- clients.zip(granted)) {
+    grantFn(client, grant)
+  }
+}
+```
+
+You may notice this looks quite similar to the original `PriorityArbiter` in its implementation.
+It uses two parameter lists in order to help the Scala type inferencer derive the types of the functions of the type of the client--we could do this with one parameter list but then it would require explicitly passing the type of the client.
```

Comment:
```suggestion
> NOTE: It uses two parameter lists in order to help the Scala type inferencer derive the types of the functions of the type of the client--we could do this with one parameter list but then it would require explicitly passing the type of the client.
```

maybe? this is sort of an aside